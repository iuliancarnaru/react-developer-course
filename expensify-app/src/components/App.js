import React, { Fragment } from "react";
import styled from "styled-components";
import GlobalStyle from "../style/GlobalStyle";
import { Router, Link, navigate } from "@reach/router";

const App = () => {
  return (
    <Fragment>
      <GlobalStyle />
      <Header />
      <Router>
        <ExpenseDashboardPage path="/" />
        <AddExpensePage path="/create" />
        <EditExpensePage path="/edit" />
        <HelpPage path="/help" />
        <NotFoundPage default />
      </Router>
    </Fragment>
  );
};

const NavLink = props => (
  <Link
    {...props}
    getProps={({ isCurrent }) => {
      return {
        style: {
          color: isCurrent ? "red" : "blue"
        }
      };
    }}
  />
);

const Header = () => {
  return (
    <header>
      <h1>Expensify</h1>
      <NavLink to="/create">Create</NavLink>
      <NavLink to="/edit">Edit</NavLink>
      <NavLink to="/help">Help</NavLink>
    </header>
  );
};

const ExpenseDashboardPage = () => {
  return (
    <div>
      <p>ExpenseDashboardPage</p>
    </div>
  );
};

const AddExpensePage = () => {
  return (
    <div>
      <p>AddExpensePage</p>
    </div>
  );
};

const EditExpensePage = () => {
  return (
    <div>
      <p>EditExpensePage</p>
    </div>
  );
};

const HelpPage = () => {
  return (
    <div>
      <p>HelpPage</p>
    </div>
  );
};

const NotFoundPage = () => {
  return (
    <div>
      <p>Uppsss ... 404 Page not found! </p>
      <Link to="/">Go back to home page.</Link>
      {/* {setTimeout(() => navigate(`/`), 2000)} */}
    </div>
  );
};

export default App;
