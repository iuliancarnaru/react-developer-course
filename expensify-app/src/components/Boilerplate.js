import React, { Fragment } from "react";
import styled from "styled-components";
import GlobalStyle from "../style/GlobalStyle";

function Boilerplate() {
  return (
    <Fragment>
      <GlobalStyle />
      <Container color="green">
        <HeadingPrimary>Boilerplate</HeadingPrimary>
        <HeadingParagraph>
          Create something beautiful with React, Reach Router and Styled
          Components
        </HeadingParagraph>
      </Container>
    </Fragment>
  );
}

const Container = styled.div`
  background-color: ${({ color }) => color || "navy"};
  padding: 2rem;
`;

const HeadingPrimary = styled.h1`
  color: yellow;
  margin: 0;
`;

const HeadingParagraph = styled.p`
  color: white;
`;

export default Boilerplate;
