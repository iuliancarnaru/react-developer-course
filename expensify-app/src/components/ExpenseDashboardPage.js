import React from "react";

const ExpenseDashboardPage = () => {
  return (
    <div>
      <p>ExpenseDashboardPage</p>
    </div>
  );
};

export default ExpenseDashboardPage;
