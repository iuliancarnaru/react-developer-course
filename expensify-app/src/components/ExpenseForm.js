import React, { Component } from "react";
import moment from "moment";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";
import { SingleDatePicker } from "react-dates";
import uuid from "uuid";

class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  handleInputChange = event => {
    const { name, value } = event.target;
    if (name !== "amount") {
      this.setState(() => ({ [name]: value }));
    } else if (!value || value.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ [name]: value }));
    }
  };

  handleSubmit = event => {
    event.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide description and amount."
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            name="description"
            id="description"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.handleInputChange}
          />
          <input
            type="text"
            name="amount"
            id="amount"
            value={this.state.amount}
            placeholder="Amount"
            onChange={this.handleInputChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={createdAt =>
              createdAt ? this.setState({ createdAt }) : ""
            }
            focused={this.state.focused}
            onFocusChange={({ focused }) => this.setState({ focused })}
            id={uuid()}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            name="note"
            id="note"
            cols="30"
            rows="10"
            placeholder="Add a note for your expense (optional)"
            onChange={this.handleInputChange}
          ></textarea>
          <button type="submit">Add expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
