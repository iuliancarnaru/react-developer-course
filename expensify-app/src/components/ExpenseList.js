import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../redux/selectors/expenses";

const ExpenseList = ({ expenses }) => (
  <div>
    <h1>Expense List</h1>
    {expenses.map(expense => (
      <ExpenseListItem key={expense.id} {...expense} />
    ))}
  </div>
);

const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
