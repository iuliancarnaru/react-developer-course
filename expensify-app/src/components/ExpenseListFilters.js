import React, { Component } from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from "../redux/actions/filters";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";
import { DateRangePicker } from "react-dates";
import uuid from "uuid";

// a controled input is when the value is controlled by JS.
class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={event => {
            this.props.dispatch(setTextFilter(event.target.value));
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={event => {
            if (event.target.value === "date") {
              this.props.dispatch(sortByDate());
            } else if (event.target.value === "amount") {
              this.props.dispatch(sortByAmount());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId={uuid()}
          endDate={this.props.filters.endDate}
          endDateId={uuid()}
          onDatesChange={({ startDate, endDate }) => {
            this.props.dispatch(setStartDate(startDate));
            this.props.dispatch(setEndDate(endDate));
          }}
          focusedInput={this.state.calendarFocused}
          onFocusChange={calendarFocused => this.setState({ calendarFocused })}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

export default connect(mapStateToProps)(ExpenseListFilters);
