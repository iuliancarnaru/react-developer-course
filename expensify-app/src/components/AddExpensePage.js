import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../redux/actions/expenses";
import { navigate } from "@reach/router";

const AddExpensePage = ({ dispatch }) => {
  return (
    <div>
      <h1>Add expense</h1>
      <ExpenseForm
        onSubmit={expense => {
          dispatch(addExpense(expense));
          setTimeout(() => navigate("/"), 0);
        }}
      />
    </div>
  );
};

export default connect()(AddExpensePage);
