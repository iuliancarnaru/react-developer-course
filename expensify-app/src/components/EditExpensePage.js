import React, { Component } from "react";
import { connect } from "react-redux";
import { navigate } from "@reach/router";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../redux/actions/expenses";

export class EditExpensePage extends Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    setTimeout(() => navigate("/"), 0);
  };

  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    setTimeout(() => navigate("/"), 0);
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: data => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
