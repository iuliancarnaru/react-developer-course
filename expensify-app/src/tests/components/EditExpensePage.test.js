import React from "react";
import { shallow } from "enzyme";
import { expenses } from "../fixtures/expenses";
import { EditExpensePage } from "../../components/EditExpensePage";

let editExpense, removeExpense, wrapper;

beforeEach(() => {
  (editExpense = jest.fn()), (removeExpense = jest.fn());
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      expense={expenses[2]}
    />
  );
});

test("should render EditExpansePage", () => {
  expect(wrapper).toMatchSnapshot();
});
test("should handle editExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[2]);
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});
test("should handle removeExpense", () => {
  wrapper.find("button").simulate("click");
  expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[2].id });
});
