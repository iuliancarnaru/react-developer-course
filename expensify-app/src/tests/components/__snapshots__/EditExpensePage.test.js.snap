// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpansePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 3500,
        "createdAt": 345600000,
        "description": "New jeans",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
